runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -1676,38 +1675,114 @@ TR_OpaqueClassBlock *
 TR_J9SharedCacheServerVM::getClassFromSignature(const char * sig, int32_t sigLength, TR_ResolvedMethod * method, bool isVettedForAOT)
    {
    TR_ResolvedRelocatableJ9JITaaSServerMethod* resolvedJITaaSMethod = (TR_ResolvedRelocatableJ9JITaaSServerMethod *)method;
-   return getClassFromSignature(sig, sigLength, (TR_OpaqueMethodBlock *)resolvedJITaaSMethod->getPersistentIdentifier(), isVettedForAOT);
+   TR_OpaqueClassBlock* clazz = NULL;
+   J9ClassLoader * cl = ((TR_ResolvedJ9Method *)method)->getClassLoader();
+   // If the classloader is the systemClassLoader, which cannot be unloaded,
+   // then look into the global, 'per client', cache
+   void * systemClassLoader = getSystemClassLoader();
+   if (cl == (J9ClassLoader*)systemClassLoader)
+      {
+      PersistentUnorderedMap<std::string, TR_OpaqueClassBlock*> & systemClassByNameMap = _compInfoPT->getClientData()->getSystemClassByNameMap();
+         {
+         OMR::CriticalSection classFromSigCS(_compInfoPT->getClientData()->getSystemClassMapMonitor());
+         auto it = systemClassByNameMap.find(std::string(sig, sigLength));
+         if (it != systemClassByNameMap.end())
+            clazz = it->second;
+         }
+      if (!clazz)
+         {
+         // classname not found; ask the client and cache the answer
+         clazz = TR_J9ServerVM::getClassFromSignature(sig, sigLength, (TR_OpaqueMethodBlock *)resolvedJITaaSMethod->getPersistentIdentifier(), isVettedForAOT);
+         if (clazz)
+            {
```

Comment:
If the client sent us back a class we must store it in the cache irrespective of the validation result. This is to catch non-AOT requests that do not need a validation.